robot/web/stream/server.go

```diff
@@ -9,120 +11,62 @@ import (
-	ServiceServer() streampb.StreamServiceServer
-
-	// NewStream creates a new stream from config and adds it for new connections to see.
-	// Returns the added stream if it is successfully added to the server.
-	NewStream(config StreamConfig) (Stream, error)
-
-	// AddStream adds the given stream for new connections to see.
-	AddStream(stream Stream) error
+type Server struct {
+	streampb.UnimplementedStreamServiceServer
 
-	// Close closes the server.
-	Close() error
+	mu                      sync.RWMutex
+	streams                 []*streamState
+	nameToStream            map[string]gostream.Stream
+	activePeerStreams       map[*webrtc.PeerConnection]map[string]*peerState
+	activeBackgroundWorkers sync.WaitGroup
 }
 
 // NewStreamServer returns a server that will run on the given port and initially starts
 // with the given stream.
-func NewStreamServer(streams ...Stream) (StreamServer, error) {
-	ss := &streamServer{
-		nameToStream:      map[string]Stream{},
+func NewServer(streams ...gostream.Stream) (*Server, error) {
+	ss := &Server{
+		nameToStream:      map[string]gostream.Stream{},
 		activePeerStreams: map[*webrtc.PeerConnection]map[string]*peerState{},
 	}
 	ss.mu.Lock()
 	defer ss.mu.Unlock()
 	for _, stream := range streams {
-		if err := ss.addStream(stream); err != nil {
+		if err := ss.add(stream); err != nil {
 			return nil, err
 		}
 	}
 	return ss, nil
 }
 
-type streamState struct {
-	mu          sync.Mutex
-	stream      Stream
-	activePeers int
-}
-
-func (ss *streamState) Start() {
-	ss.mu.Lock()
-	defer ss.mu.Unlock()
-	ss.activePeers++
-	if ss.activePeers == 1 {
-		ss.stream.Start()
-	}
-}
-
-func (ss *streamState) Stop() {
-	ss.mu.Lock()
-	defer ss.mu.Unlock()
-	ss.activePeers--
-	if ss.activePeers <= 0 {
-		ss.activePeers = 0
-		ss.stream.Stop()
-	}
-}
-
-type peerState struct {
-	stream  *streamState
-	senders []*webrtc.RTPSender
-}
-
-type streamServer struct {
-	mu                      sync.RWMutex
-	streams                 []*streamState
-	nameToStream            map[string]Stream
-	activePeerStreams       map[*webrtc.PeerConnection]map[string]*peerState
-	activeBackgroundWorkers sync.WaitGroup
-}
-
-func (ss *streamServer) ServiceServer() streampb.StreamServiceServer {
-	return &streamRPCServer{ss: ss}
-}
-
-func (ss *streamServer) NewStream(config StreamConfig) (Stream, error) {
+func (ss *Server) NewStream(config gostream.StreamConfig) (gostream.Stream, error) {
 	ss.mu.Lock()
 	defer ss.mu.Unlock()
 
 	if _, ok := ss.nameToStream[config.Name]; ok {
 		return nil, &StreamAlreadyRegisteredError{config.Name}
 	}
-	stream, err := NewStream(config)
+	stream, err := gostream.NewStream(config)
 	if err != nil {
 		return nil, err
 	}
-	if err := ss.addStream(stream); err != nil {
+	if err := ss.add(stream); err != nil {
 		return nil, err
 	}
 	return stream, nil
 }
 
-func (ss *streamServer) AddStream(stream Stream) error {
+func (ss *Server) Add(stream gostream.Stream) error {
```

Comment:
One of the bits that annoyed me about the two "stream server" types was that we had the `streamServer.AddStream` call and the `streamRPCServer.AddStream` call. And I kept wanting the latter, but searching through text kept pointing me here first.